common/json_tok.c

```diff
@@ -584,3 +584,76 @@ struct command_result *param_extra_tlvs(struct command *cmd, const char *name,
 	*fields = temp;
 	return NULL;
 }
+
+struct command_result *param_routehint(struct command *cmd, const char *name,
+				       const char *buffer, const jsmntok_t *tok,
+				       struct route_info **ri)
+{
+	size_t i;
+	const jsmntok_t *curr;
+	char *element_name;
+	const char *err;
+
+	if (tok->type != JSMN_ARRAY) {
+		return command_fail(
+		    cmd, JSONRPC2_INVALID_PARAMS,
+		    "Routehint %s (\"%s\") is not an array of hop objects",
+		    name, json_strdup(tmpctx, buffer, tok));
+	}
+
+	*ri = tal_arr(cmd, struct route_info, 0);
+	json_for_each_arr(i, curr, tok) {
+		struct route_info e;
+		struct amount_msat temp;
+		element_name = tal_fmt(cmd, "%s[%zu]", name, i);
+
+		err = json_scan(tmpctx, buffer, curr,
+				"{id:%,scid:%,feebase:%,feeprop:%,expirydelta:%}",
+				JSON_SCAN(json_to_node_id, &e.pubkey),
+				JSON_SCAN(json_to_short_channel_id, &e.short_channel_id),
+				JSON_SCAN(json_to_msat, &temp),
+				JSON_SCAN(json_to_u32, &e.fee_proportional_millionths),
+				JSON_SCAN(json_to_u16, &e.cltv_expiry_delta)
+			);
+		e.fee_base_msat =
+		    temp.millisatoshis; /* Raw: internal conversion. */
+		if (err != NULL) {
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Error parsing routehint %s: %s",
+					    element_name, err);
+		}
+		tal_arr_expand(ri, e);
+		tal_free(element_name);
+	}
+	return NULL;
+}
+
+struct command_result *
+param_routehint_array(struct command *cmd, const char *name, const char *buffer,
+		      const jsmntok_t *tok, struct route_info ***ris)
+{
+	size_t i;
+	const jsmntok_t *curr;
+	char *element_name;
+	struct command_result *err;
+	if (tok->type != JSMN_ARRAY) {
+		return command_fail(
+		    cmd, JSONRPC2_INVALID_PARAMS,
+		    "Routehint array %s (\"%s\") is not an array",
+		    name, json_strdup(tmpctx, buffer, tok));
+	}
+
+	*ris = tal_arr(cmd, struct route_info *, 0);
```

Comment:
Here too, tok->size?

Reply:
I tried this but there are just too many levels of indirection and my head started hurting: we're passing a pointer of position in an array, which should then be initialized in the callee, and is itself a variable length array. What would the canonical way of doing this be? And is there a trick to make this clearer in the code (parentheses, ...)?